lib/ansible/plugins/connection/qubes.py

```diff
@@ -0,0 +1,186 @@
+
+        display.vvv("RUN %s" % (local_cmd,), host=self._remote_vmname)
+        p = subprocess.Popen(local_cmd, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # Here we are writing the actual command to the remote bash
+        p.stdin.write(to_bytes("%s" % cmd, errors='surrogate_or_strict'))
+        stdout, stderr = p.communicate(input=in_data)
+        stdout = to_bytes(stdout, errors='surrogate_or_strict')
+        stderr = to_bytes(stderr, errors='surrogate_or_strict')
+        return p.returncode, stdout, stderr
+
+    def _connect(self):
+        """
+        No persistent connection is being maintained.
+        """
+        super(Connection, self)._connect()
+        self._connected = True
+
+    @ensure_connect
+    def exec_command(self, cmd, in_data=None, sudoable=False):
+        """ run specified command in a running QubesVM """
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.vvvv("CMD IS: %s" % cmd)
+
+        rc, stdout, stderr = self._qubes(cmd)
+
+        display.vvvvv("STDOUT %r STDERR %r" % (stderr, stderr))
+        return rc, stdout, stderr
+
+    def put_file(self, in_path, out_path):
+        """ Place a local file located in 'in_path' inside container at 'out_path' """
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self._remote_vmname)
+
+        cmd = "qvm-copy-to-vm " + self._remote_vmname + " " + in_path
+        cmd_args_list = shlex.split(to_native(cmd, errors='surrogate_or_strict'))
+
+        p = subprocess.Popen(cmd_args_list, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        stdout, stderr = p.communicate()
+
+        # Now let us move the file to the right location
+        filename = os.path.basename(in_path)
+        cmd = "mv {0} {1}".format(os.path.join("/home/{0}/QubesIncoming".format(self.user), self._hostname,
+                                               filename), out_path)
+
+        self._qubes(cmd)
+
+    def fetch_file(self, in_path, out_path):
+        """ obtain file specified via 'in_path' from the container and place it at 'out_path' """
+        super(Connection, self).fetch_file(in_path, out_path)
+        display.vvv("FETCH %s TO %s" % (in_path, out_path), host=self._remote_vmname)
+
+        cmd = ""
+        # Find the actual filename
+        filename = os.path.basename(in_path)
+
+        # We are running in dom0
+        cmd_args_list = ["qvm-run", "--pass-io", self._remote_vmname, "'cat {0}'".format(in_path)]
+        p = subprocess.Popen(cmd_args_list, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        stdout, stderr = p.communicate()
+        tmp_path = os.path.join(C.DEFAULT_LOCAL_TMP, filename)
```

Comment:
Should we be using a randomized temp filename?  something like: ```tempfile.mkstemp([..], dir=C.DEFAULT_LOCAL_TMP)``` ?

Reply:
I did not want to do that as this always over writes whatever is there.

Reply:
<nod> But it seems like it's vulnerable to both symlink attacks and denial of service attacks unless it's randomized. (if it is randomized, it is resistant to but not immune to DOS.  Using mkstemp should solve symlink attacks.)

Reply:
No longer using a temporary file.  This issue is resolved